### DISCUSSION TITLE
Multiple Sites for Bangs!

### DISCUSSION THREAD

**Discussion by cfuendev** [minbrowser/min#1795] [Ideas]
Is there a plan to support the option to launch multiple sites with a single bang? It'd be nice to type something like "!d" (For dashboard) and get my email, twitter, my uni's moodle, whasapp and other sites launched on a single keypress.

If this is a completely new idea for the Min team and you hadn't thought about it until now, I'd like to be the one to try and implement it! To save you some work (And for the dev experience).

**Comment by PalmerAL** [comment]
Yeah, that's a good idea, go for it. Code is around here: https://github.com/minbrowser/min/blob/7749a35ea71a5f373c05e1d587b620a7a5f7c1fc/js/searchbar/bangsPlugin.js#L224

I think the easiest thing to do would be to let the user type multiple URLs in the custom bang URL input, and then split the text and open each URL when the bang is entered. For keyboard shortcuts, we allow multiple entries with a comma between them, but it may be possible for URLs to have a comma in them as well, which would be an issue here. Maybe a less-common separator would work?

Also, hello fellow IEEE member :)

**Reply by cfuendev** [reply]
Hey! I finally have the time in my hands to work on this. I forked the repo and I'm analyzing the whole project a bit and taking notes before touching the code. I'll try to finish this as soon as possible since I took a long time to get all my university work together before taking care of this.

I had no idea urls could hold commas. One thing I could do is perform a regex operation that splits the multiple URLs by the last character (Which would be the comma) to make sure it's not splitting a single URL into a disfigured mess. But I'll do take the idea of using an obscure separator into account.

Also, Hello! It's nice to get to know an IEEE fellow of a different place :^)

**Reply by PalmerAL** [reply]
Sounds good! There's no particular rush, take your time.

I'm not sure I completely understand the regex idea. We don't require URLs to begin with a protocol, so this works:
<img width="748" alt="Screen Shot 2021-12-24 at 12 26 53 PM" src="https://user-images.githubusercontent.com/10314059/147368740-8eaad1fc-5f61-460f-b0c7-c1e36711c605.png">

Which is good to have I think, since not everyone will remember to type a protocol.

But then if you enter something like this:
> http://example.com/test,example.com,example.com

It could be 1, 2, or 3 URLs depending on how you split it.

**Reply by cfuendev** [reply]
> I'm not sure I completely understand the regex idea.

I was trying to implement an expression that would identify the comma that separated an URL from another, but after some knowledge acquisition through YouTube crash courses and re-thinking the idea for a bit, not only have I realized how impossible it's to tell the correct comma just using a regular expression, but realistically people are just gonna get confused by trying to separate URLs that already have commas in them **by using more commas**.

I came up with 
```javascript
/(?<=,.*?),(?=.*?\.)/g
```
However, it started failing as soon as one of the URLs had more than a single comma in it lol
For example, the links below would get incorrectly split.
```javascript
"http://example.com/?foo=1,bar=2,pepe=false,google.com,https://website.com,nanana"
```

That's when I took a step back and realized, even if I got it to work, it would be pretty hard to stare at. Just look at that string. XD People can get confused when trying to embed multiple sites into a single bang. Specially if they're not tech-savy enough. So I think I'm going with the obscure operator approach you suggested.

> Maybe a less-common separator would work?

After a tad of research, I think a pipe operator `|` makes sense. Not only is it synonymous with logical "OR", which is pretty intuitive, and pretty large, which will make it easy to tell websites appart, but it appears to me it isn't very used, and if ever used, it needs to be already encoded as `%7C`. Even in [the original RFC specification for URIs](https://www.ietf.org/rfc/rfc2396.txt), pipes are only used to show alternative syntax (And ascii art), but it's never specified as a component for URI queries and the like.

So I think I will implement the multiple site !bangs with that symbol. :^)

**Reply by PalmerAL** [reply]
That all makes sense to me; the pipe separator seems like a good solution.

Anecdotally, I searched through my browsing history in Min to see how often either of these get used. Out of 3596 entries, 71 had commas (although most of those were google maps links). Only 3 entries had the encoded pipe operator, and 0 had an unencoded one.
